src/service_data_structure.cc

```diff
@@ -0,0 +1,531 @@
+        ok = chirp_connect_backend::GetChirp(chirp_id, &chirp);
+        CHECK(ok) << "The chirp with chirp_id `" << chirp_id << "` should exist.";
+
+        if (timercmp(&(chirp.time), from, >=) && timercmp(&(chirp.time), &now, <)) {
+          ret.insert(chirp_id);
+        }
+      }
+    }
+  }
+
+  *from = now;
+  return ret;
+}
+
+ServiceDataStructure::ReturnCodes ServiceDataStructure::UserRegister(const std::string &username) {
+  // Invalid username
+  if (username.empty()) {
+    return INVALID_ARGUMENT;
+  }
+
+  bool user_found = chirp_connect_backend::GetUser(username, nullptr);
+  // If the specified username is not found
+  if (!user_found) {
+    struct User new_user(username);
+    UserChirpList chirp_list;
+    UserFollowingList following_list;
+    bool ok = chirp_connect_backend::SaveUser(username, new_user) &&
+              chirp_connect_backend::SaveUserChirpList(username, chirp_list) &&
+              chirp_connect_backend::SaveUserFollowingList(username, following_list);
+    if (!ok) {
+      chirp_connect_backend::DeleteUser(username);
+      chirp_connect_backend::DeleteUserChirpList(username);
+      chirp_connect_backend::DeleteUserFollowingList(username);
+      return INTENRAL_BACKEND_ERROR;
+    } else {
+      return OK;
+    }
+  } else {
+    return USER_EXISTS;
+  }
+
+  // Should not enter this line
+  return UNKOWN_ERROR;
+}
+
+std::unique_ptr<ServiceDataStructure::UserSession> ServiceDataStructure::UserLogin(const std::string &username) {
+  bool user_found = chirp_connect_backend::GetUser(username, nullptr);
+
+  // If the specified username is found
+  if (user_found) {
+    return std::unique_ptr<ServiceDataStructure::UserSession>(new UserSession(username));
+  } else {
+    return nullptr;
+  }
+}
+
+ServiceDataStructure::User::User(const std::string &username) : username(username) {
+  gettimeofday(&last_update_chirp_time, nullptr);
+}
+
+ServiceDataStructure::Chirp::Chirp(const std::string &user, const uint64_t &parent_id, const std::string &text)
+    : user(user), parent_id(parent_id), text(text) {
+  id = chirp_connect_backend::GetNextChirpId();
+  gettimeofday(&time, nullptr);
+}
+
+// Type identifier for serializing data
+const std::string kTypeNextChirpId({0, 0, 0, char(1)});
+const std::string kTypeUsernameToUserPrefix({0, 0, 0, char(2)});
+const std::string kTypeUsernameToFollowingPrefix({0, 0, 0, char(3)});
+const std::string kTypeUsernameToChirpPrefix({0, 0, 0, char(4)});
+const std::string kTypeChirpidToChirpPrefix({0, 0, 0, char(5)});
+
+void chirp_connect_backend::DecomposeBinaryUser(const std::string &input, struct ServiceDataStructure::User * const user) {
+  if (user != nullptr) {
+    user->last_update_chirp_time = *(reinterpret_cast<const struct timeval *>(input.c_str()));
+    user->username = input.substr(sizeof(struct timeval));
+  }
+}
+
+void chirp_connect_backend::ComposeBinaryUser(const struct ServiceDataStructure::User &user, std::string * const output) {
```

Comment:
for the output arg the ordering is confusing. would `const std::string *output` work?

Reply:
No.

Reply:
are you trying to get a pointer to a const string?